pkg/kubectl/cmd/plugin.go

```diff
@@ -18,170 +18,172 @@ package cmd
+				pluginWarnings++
 			}
-		},
+		}
 	}
 
-	for _, flag := range plugin.Flags {
-		cmd.Flags().StringP(flag.Name, flag.Shorthand, flag.DefValue, flag.Desc)
+	if !pluginsFound {
+		return fmt.Errorf("error: unable to find any kubectl plugins in your PATH")
 	}
 
-	for _, childPlugin := range plugin.Tree {
-		cmd.AddCommand(NewCmdForPlugin(f, childPlugin, runner, streams))
+	if pluginWarnings > 0 {
+		fmt.Fprintln(o.ErrOut)
+		if pluginWarnings == 1 {
+			return fmt.Errorf("one plugin warning was found")
+		}
+		return fmt.Errorf("%v plugin warnings were found", pluginWarnings)
 	}
 
-	return cmd
-}
-
-type flagsPluginEnvProvider struct {
-	cmd *cobra.Command
+	return nil
 }
 
-func (p *flagsPluginEnvProvider) Env() (plugins.EnvList, error) {
-	globalPrefix := "KUBECTL_PLUGINS_GLOBAL_FLAG_"
-	env := plugins.EnvList{}
-	p.cmd.InheritedFlags().VisitAll(func(flag *pflag.Flag) {
-		env = append(env, plugins.FlagToEnv(flag, globalPrefix))
-	})
-	localPrefix := "KUBECTL_PLUGINS_LOCAL_FLAG_"
-	p.cmd.LocalFlags().VisitAll(func(flag *pflag.Flag) {
-		env = append(env, plugins.FlagToEnv(flag, localPrefix))
-	})
-	return env, nil
+// pathVerifier receives a path and determines if it is valid or not
+type PathVerifier interface {
+	// Verify determines if a given path is valid
+	Verify(path string) error
 }
 
-type factoryAttrsPluginEnvProvider struct {
-	factory cmdutil.Factory
+type CommandOverrideVerifier struct {
+	root *cobra.Command
 }
 
-func (p *factoryAttrsPluginEnvProvider) Env() (plugins.EnvList, error) {
-	cmdNamespace, _, err := p.factory.ToRawKubeConfigLoader().Namespace()
-	if err != nil {
-		return plugins.EnvList{}, err
+// Verify implements PathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kubectl command path.
+func (v *CommandOverrideVerifier) Verify(path string) error {
+	if v.root == nil {
+		return fmt.Errorf("unable to verify path with nil root")
 	}
-	return plugins.EnvList{
-		plugins.Env{N: "KUBECTL_PLUGINS_CURRENT_NAMESPACE", V: cmdNamespace},
-	}, nil
-}
 
-// pluginLoader loads plugins from a path set by the KUBECTL_PLUGINS_PATH env var.
-// If this env var is not set, it defaults to
-//   "~/.kube/plugins", plus
-//  "./kubectl/plugins" directory under the "data dir" directory specified by the XDG
-// system directory structure spec for the given platform.
-func pluginLoader() plugins.PluginLoader {
-	if len(os.Getenv("KUBECTL_PLUGINS_PATH")) > 0 {
-		return plugins.KubectlPluginsPathPluginLoader()
+	// extract the plugin binary name
+	segs := strings.Split(path, "/")
+	binName := segs[len(segs)-1]
+
+	cmdPath := strings.Split(binName, "-")
+	if len(cmdPath) > 1 {
+		// the first argument is always "kubectl" for a plugin binary
+		cmdPath = cmdPath[1:]
 	}
-	return plugins.TolerantMultiPluginLoader{
-		plugins.XDGDataDirsPluginLoader(),
-		plugins.UserDirPluginLoader(),
+
+	if cmd, _, err := v.root.Find(cmdPath); err == nil {
```

Comment:
Shouldn't you be also checking if the command is executable?

Reply:
Thanks, added this as part of the "Verify" method